mongodb-atlas/src/main/java/io/vantiq/atlasConnector/AtlasStorageMgr.java

```diff
@@ -535,26 +537,29 @@ private Object cleanQualMap(Object qual) {
      */
     <T> Flowable<T> usingSession(String databaseName, String xid, BiFunction<MongoDatabase, ClientSession,
                                  Publisher<T>> cmdFunction) {
-        Single<ImmutablePair<MongoDatabase, ClientSession>> sessionDbObs;
-        sessionDbObs = sessions.asMap().computeIfAbsent(databaseName, name -> {
-            Single<MongoClient> clientObs = connection.connect(config);
-            return clientObs.flatMapPublisher(client -> {
-                Flowable<ClientSession> sessionObs = Flowable.fromPublisher(client.startSession());
-                return sessionObs.map(session -> new ImmutablePair<>(client.getDatabase(databaseName), session));
-            }).compose(new CacheIfNotEmptyOrError<>()).firstOrError();
-        });
+        Single<ImmutablePair<MongoDatabase, ClientSession>> dbSessionObs;
         if (xid != null) {
-            // if we are in a transaction substitute its session to use instead of the default from above
-            sessionDbObs = sessionDbObs.flatMap(sessionDb -> transactions.asMap().getOrDefault(xid,
-                    Single.error(new Exception("No transaction found for id " + xid))).map(session ->
-                new ImmutablePair<>(sessionDb.getLeft(), session)
-            ));
+            // if we are in a transaction use its client and session to run the command
+            val errorIfMissing = Single.<ImmutablePair<MongoClient, ClientSession>>error(new Exception(
+                    "No transaction found for id " + xid));
+            dbSessionObs = transactions.asMap().getOrDefault(xid, errorIfMissing).observeOn(rxScheduler).map(cliSess ->
```

Comment:
cannot mix the transaction session with a client that did not create it. was getting away with this before b/c every request was shunted onto a single event loop thread